File: src/main/scala/li/cil/oc/common/capabilities/CapabilityEnvironment.java
package li.cil.oc.common.capabilities;

import li.cil.oc.api.util.Location;
import li.cil.oc.api.network.NodeContainer;
import li.cil.oc.api.network.Message;
import li.cil.oc.api.network.Node;
import net.minecraft.nbt.NBTBase;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.EnumFacing;
import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;
import net.minecraftforge.common.capabilities.CapabilityManager;

import javax.annotation.Nullable;

public final class CapabilityEnvironment {
    @CapabilityInject(NodeContainer.class)
    public static Capability<NodeContainer> ENVIRONMENT_CAPABILITY;

    public static void register() {
        CapabilityManager.INSTANCE.register(NodeContainer.class, new Capability.IStorage<NodeContainer>() {
            @Nullable
            @Override
            public NBTBase writeNBT(final Capability<NodeContainer> capability, final NodeContainer instance, final EnumFacing side) {
                return null;
            }

            @Override
            public void readNBT(final Capability<NodeContainer> capability, final NodeContainer instance, final EnumFacing side, final NBTBase nbt) {
            }
        }, () -> new NodeContainer() {
            @Override
            public Location getLocation() {
                return null;
            }

            @Override
            public Node getNode() {
                return null;
            }

            @Override
            public void onConnect(final Node node) {
            }

            @Override
            public void onDisconnect(final Node node) {
            }

            @Override
            public void onMessage(final Message message) {
            }

            @Override
            public boolean canConnect() {
                return false;
            }

            @Override
            public NBTTagCompound serializeNBT() {
                return new NBTTagCompound();
            }

            @Override
            public void deserializeNBT(final NBTTagCompound nbt) {
            }
        });
    }

    private CapabilityEnvironment() {
    }
}

File: src/main/scala/li/cil/oc/common/capabilities/CapabilityEnvironment.scala
package li.cil.oc.common.capabilities

import li.cil.oc.api
import li.cil.oc.api.network.Environment
import li.cil.oc.api.network.Message
import li.cil.oc.api.network.Node
import li.cil.oc.api.network.Visibility
import li.cil.oc.integration.Mods
import net.minecraft.nbt.INBT
import net.minecraft.nbt.CompoundNBT
import net.minecraft.tileentity.TileEntity
import net.minecraft.util.Direction
import net.minecraft.util.ResourceLocation
import net.minecraftforge.common.capabilities.Capability
import net.minecraftforge.common.capabilities.ICapabilityProvider
import net.minecraftforge.common.util.LazyOptional
import net.minecraftforge.common.util.NonNullSupplier

object CapabilityEnvironment {
  final val ProviderEnvironment = new ResourceLocation(Mods.IDs.OpenComputers, "environment")

  class Provider(val tileEntity: TileEntity with Environment) extends ICapabilityProvider with NonNullSupplier[Provider] with Environment {
    private val wrapper = LazyOptional.of(this)

    def get = this

    def invalidate() = wrapper.invalidate

    override def getCapability[T](capability: Capability[T], facing: Direction): LazyOptional[T] = {
      if (capability == Capabilities.EnvironmentCapability) wrapper.cast[T]
      else LazyOptional.empty[T]
    }

    override def node = tileEntity.node

    override def onMessage(message: Message) = tileEntity.onMessage(message)

    override def onConnect(node: Node) = tileEntity.onConnect(node)

    override def onDisconnect(node: Node) = tileEntity.onDisconnect(node)
  }

  class DefaultImpl extends Environment {
    override val node = api.Network.newNode(this, Visibility.None).create()

    override def onMessage(message: Message): Unit = {}

    override def onConnect(node: Node): Unit = {}

    override def onDisconnect(node: Node): Unit = {}
  }

  class DefaultStorage extends Capability.IStorage[Environment] {
    override def writeNBT(capability: Capability[Environment], t: Environment, facing: Direction): INBT = {
      val node = t.node
      if (node != null) {
        val nbt = new CompoundNBT()
        node.saveData(nbt)
        nbt
      }
      else null
    }

    override def readNBT(capability: Capability[Environment], t: Environment, facing: Direction, nbtBase: INBT): Unit = {
      nbtBase match {
        case nbt: CompoundNBT =>
          val node = t.node
          if (node != null) node.loadData(nbt)
        case _ =>
      }
    }
  }

}
